File: python/resistor-color-duo/resistor_color_duo.py
""" https://exercism.org/tracks/python/exercises/resistor-color-duo """

COLOR_CODES = {
    "black": 0,
    "brown": 1,
    "red": 2,
    "orange": 3,
    "yellow": 4,
    "green": 5,
    "blue": 6,
    "violet": 7,
    "grey": 8,
    "white": 9,
}


def value(colors: list[str]) -> int:
    """
    Returns the color code of a given color, or two-color combination
    ex: "brown" would return 1; "brown-green" would return 15

    Args:
        colors: list of color strings

    Returns:
        int: corresponding color code
    """

    if not isinstance(colors, list) or not all(
        isinstance(color, str) for color in colors
    ):
        raise TypeError("colors must be list of color strings")

    output = [
        str(COLOR_CODES[color.lower()])
        for color in colors
        if color.lower() in COLOR_CODES
    ][:2]

    return int("".join(output))

File: typescript/resistor-color-duo/resistor-color-duo.ts
/**
 * Array of resistor color names in order of their numeric values (0-9)
 */
export const COLORS = [
  "black",
  "brown",
  "red",
  "orange",
  "yellow",
  "green",
  "blue",
  "violet",
  "grey",
  "white",
]

/**
 * Gets the numeric value for a resistor color
 * @param color - The color name
 * @returns The numeric value (0-9) for the color
 */
const colorCode = (color: string): number => {
  const index = COLORS.indexOf(color.toLowerCase())
  if (index === -1) {
    throw new Error(`Invalid color: ${color}`)
  }
  return index
}

/**
 * Decodes the resistance value from the first two color bands
 * @param colors - Array of color names (only first two are used)
 * @returns Two-digit number representing the resistance value
 */
export function decodedValue(colors: string[]): number {
  if (colors.length < 2) {
    throw new Error("At least two colors are required")
  }

  const firstDigit = colorCode(colors[0])
  const secondDigit = colorCode(colors[1])

  return firstDigit * 10 + secondDigit
}
